vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,314 @@
+        Map config = ConfigurationHelper.newInstance(this)
+                                        .loadStepDefaults()
+                                        .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+                                        .mixinStageConfig(script.commonPipelineEnvironment, arguments.stageName ?: env.STAGE_NAME, STEP_CONFIG_KEYS)
+                                        .mixin(arguments, PARAMETER_KEYS)
+                                        .use()
+
+        String manifestFilePath = config.manifestFile ?: "manifest.yml"
+        String variablesFilePath = config.variablesFile ?: "manifest-variables.yml"
+        String outputFilePath = config.outputManifestFile ?: manifestFilePath
+
+        // Parameter for tests only. No public API.
+        // This is to skip manifest file deletion in tests, which
+        // would be a nuisance, as it alters the sources repository.
+        Boolean skipDeletion = arguments.skipDeletion ?: false
+
+        File manifestFile = new File(manifestFilePath)
+        File variablesFile = new File(variablesFilePath)
+
+        if (!manifestFile.exists()) {
+            echo "[YamlSubstituteVariables] Could not find YAML file at ${manifestFilePath}. Skipping variable substitution."
+            return
+        }
+
+        if (!variablesFile.exists()) {
+            echo "[YamlSubstituteVariables] Could not find variable substitution file at ${variablesFilePath}. Skipping variable substitution."
+            return
+        }
+
+        def manifestData = null;
+        try {
+            // may return a List<Object>  (in case more YAML segments are in the file)
+            // or a Map<String, Object> in case there is just one segment.
+            manifestData = readYaml file: manifestFilePath
+            echo "[YamlSubstituteVariables] Loaded manifest at ${manifestFilePath}!"
+        }
+        catch(Exception ex) {
+            debug("Exception: ${ex}")
+            error "[YamlSubstituteVariables] Could not load manifest file at ${manifestFilePath}."
+            error "[YamlSubstituteVariables] Exception was: ${ex}"
+        }
+
+        def variablesData = null
+        try {
+            // may return a List<Object>  (in case more YAML segments are in the file)
+            // or a Map<String, Object> in case there is just one segment.
+            variablesData = readYaml file: variablesFilePath
+            echo "[YamlSubstituteVariables] Loaded variables file at ${variablesFilePath}!"
+        }
+        catch(Exception ex) {
+            debug("Exception: ${ex}")
+            error "[YamlSubstituteVariables] Could not load manifest variables file at ${variablesFilePath}."
+            error "[YamlSubstituteVariables] Exception was: ${ex}"
+        }
+
+        // substitute all variables.
+        ExecutionContext context = new ExecutionContext()
+        def result = substitute(manifestData, variablesData, context)
```

Comment:
Maybe it would be better to provide the variable data already parsed rather than providing a file (or at least provide that as additional option). More use cases can be handled with that step in this case.

Reply:
Actually, both `manifestData` as well as `variablesData` are already parsed Yamls (i.e. an Object graph). The returned value of `substitute(...)` is as well an Object graph, i.e. parsed data.

Reply:
ok, my comment was misplaced. The method signature of the step (aka the map handed over to `call`) expects files. In my opinion it would be better to support already parsed data here (maybe in addition to the files, like it is now). With such an approach I can provide a memory model (which I can manipulate beforehand). Otherwise in case something needs to be manipulated on the fly I have to create the file, which needs to be parsed later on again.

Reply:
This has now been outsourced into two scripts:
* `cfManifestSubstituteVariables` which deals with files and is intended to do replacements in cf manifest files mainly (though not necessarily limited to them).
* `yamlSubstituteVariables` which deals with the variable substitution based on Yaml objects (that may or may not have been parsed from files beforehand).